src/input-states.js

```diff
@@ -0,0 +1,222 @@
+    } else if (!_.isString(vnid)) {
+        throw Error("Invalid vnid: " + vnid);
+    } else if (_.isString(portName) && !node.inPorts[portName]) {
+        throw Error("Invalid portName: " + portName);
+    } else if (arguments.length > 4) {
+        setPortStateBySocketId(node, vnid, portName, socketId, state);
+        return this;
+    } else if (_.isString(portName)) {
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    } else if (_.isObject(arguments[2])) {
+        setAllPortStates(node, vnid, arguments[2]);
+        return this;
+    } else {
+        return getAllPortStates(node, vnid);
+    }
+}
+
+/**
+ * Returns a hash of port names to port states or array of states if the port is
+ * addressable/multi. This provides the caller with all portNames, distingishes between
+ * addressable/multi and non-addressable/multi ports, and provides the caller with all their
+ * states that will be useful when calling an updater.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ */
+function getAllPortStates(node, vnid) {
+    return _.mapObject(node.inPorts, function(port, portName){
+        return getPortState(node, vnid, portName);
+    });
+}
+
+/**
+ * Returns the state of the port or an array of states, if the port is
+ * addressable/multi. When all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortState(node, vnid, portName) {
+    var port = node.inPorts[portName];
+    if (port.isMulti())
+        return getPortStateArray(node, vnid, portName);
+    else if (vniStateExists(node, vnid, portName))
+        return node.vnis[vnid].inputStates[portName];
+    else if (vnid) // By default states that come in on vnid '' apply to all
+        return getPortState(node, '', portName);
+    else
+        return undefined;
+}
+
+/**
+ * Returns an array of the port states, one item for each attached socket. When
+ * all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortStateArray(node, vnid, portName) {
+    var port = node.inPorts[portName];
+    if (!port.isMulti()) throw Error("This port is not addressable/multi: " + portName);
+    else return port.listAttached().map(function(socketId){
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    });
+}
+
+/**
+ * Returns the state of this port for the given socketId. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketId is not neccessarily the same as the array index
```

Comment:
SocketId is the array index of node.vnis[vnid].inputStates[portName], but not neccessarily the array index of getPortStateArray(node, vnid, portName). The later only includes attached sockets.